File: src/services/announcementsStore.js
/**
 * In-memory announcements store.
 * Swap this with a DB later (SQLite/Postgres/etc) without changing routes.
 */

/** @typedef {{ id: string, title: string, body: string, createdAt: string }} Announcement */

/** @type {Map<string, Announcement>} */
const announcements = new Map();

function newId() {
    return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 10)}`;
    }

    export function listAnnouncements() {
    return Array.from(announcements.values()).sort((a, b) =>
        b.createdAt.localeCompare(a.createdAt)
    );
    }

    export function createAnnouncement({ title, body }) {
    const item = {
        id: newId(),
        title,
        body,
        createdAt: new Date().toISOString()
    };
    announcements.set(item.id, item);
    return item;
    }

    export function getAnnouncement(id) {
    return announcements.get(id) ?? null;
    }

    export function updateAnnouncement(id, patch) {
    const existing = announcements.get(id);
    if (!existing) return null;

    const updated = {
        ...existing,
        ...(patch.title !== undefined ? { title: patch.title } : null),
        ...(patch.body !== undefined ? { body: patch.body } : null)
    };

    announcements.set(id, updated);
    return updated;
    }

    export function deleteAnnouncement(id) {
    return announcements.delete(id);
}

File: src/announcements_store.py
from .models import Announcement, db
from typing import Optional, Dict, Any, List

def list_announcements() -> List[Dict[str, Any]]:
    announcements = Announcement.query.order_by(Announcement.created_at.desc()).all()
    return [announcement.to_dict() for announcement in announcements]

def create_announcement(data: Dict[str, Any]) -> Dict[str, Any]:
    title = data.get("title")
    body = data.get("body")
    author_id = data.get("authorId")

    announcement = Announcement(title=title, body=body, author_id=author_id)
    db.session.add(announcement)
    db.session.commit()
    return announcement.to_dict()

def get_announcement(announcement_id: str) -> Optional[Dict[str, Any]]:
    announcement = Announcement.query.get(announcement_id)
    return announcement.to_dict() if announcement else None

def update_announcement(announcement_id: str, patch: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    announcement = Announcement.query.get(announcement_id)
    if not announcement:
        return None

    if "title" in patch:
        announcement.title = patch["title"]
    if "body" in patch:
        announcement.body = patch["body"]

    db.session.commit()
    return announcement.to_dict()

def delete_announcement(announcement_id: str) -> bool:
    announcement = Announcement.query.get(announcement_id)
    if not announcement:
        return False

    db.session.delete(announcement)
    db.session.commit()
    return True